---
title: "Analyzing genes of interest with SolveBio"
output: rmarkdown::html_document
---
In this example, we will show how SolveBio can be used to create a customized data table with specific fields and records extracted from SolveBio's datasets. We start with a list of 15 genes and we query two of SolveBio's reference datasets, ClinVar and GWAS for their variants. SolveBio enables an easy integration of diverse data sources by standardizing the keys for common genomic entitities such as genes and variants. This example also shows how to select only a subset of the reference data for download by specifying filters and fields of interest. Importantly, we show how data from two different sources can be merged into one data table by SolveBio variant ID. In the case of available expression data for the gene list in question, we also include examples showing two ways of generating gene expression heatmaps.

```{r, message = FALSE, warning=FALSE}
library(solvebio)
library(data.table)
```

login to SolveBio:

```{r eval=FALSE}
login(api_key="<api_key>", api_host="<api_host>")
```

Define a list of genes for the analysis and create a filter that will be used to query SolveBio datasets:

```{r}
# random sample of 15 genes from clinvar gene_symbols generated by running the R sample function:
genes <- c("MYBPC3", "RDH5", "MYO15A", "FZD4", "DMD", "GLIS3", "ISPD", "C3", "APOB", "ATP6V0A2", "KIF5A", "SNCA", "LDLR", "ANK1", "FBN2")
genes_filter <- list(list('gene_symbol__in', paste(genes)))
```

SolveBio allows you to query its datasets only for the gene set and fields of interest. In this example, we will query the ClinVar/Combined and GWAS datasets to extract variants and other supporting information for our sample gene list. 

We first query the ClinVar/Combined dataset for the genes in our genes_filter and specify which columns (fields) we want to export:

```{r}
# view the complete dataset at https://my.solvebio.com/data/510111974568996872
ClinVar <- Dataset.get_by_full_path("solveBio:public:/ClinVar/3.7.4-2017-01-30/Combined-GRCh37")
clinvar_genes_q <- Dataset.query(id = ClinVar$id,
                                filters=genes_filter, 
                                fields=c("variant_sbid", "phenotype", 
                                         "clinical_significance", "gene_symbol"), 
                                paginate=TRUE)

clinvar_genes_dt <- as.data.table(clinvar_genes_q)
```

Let's now query the GWAS dataset for the same gene list:

```{r}
# view the complete dataset at https://my.solvebio.com/data/510109000607280262
gwas <- Dataset.get_by_full_path("solveBio:public:/GWAS/2.1.1-2017-01-10/GWAS-GRCh37")
gwas_q <- Dataset.query(id = gwas$id, filters=genes_filter, genome_build="GRCh37", paginate=TRUE)
gwas_dt <- as.data.table(gwas_q)
```

SolveBio has enabled easy navigation between these two datasets by standardizing the formats of common entity types such as gene symbols and variant IDs. Let's merge them by their common field - 'variant_sbid'.

Before we merge with ClinVar/Combined, we need to unlist the variant_sbid column in GWAS:

```{r, echo = TRUE}
# add a column to index the rows in the gwas data table
gwas_dt$rowID <- seq.int(nrow(gwas_dt))

# add another column with the length of the list in variant_sbid
gwas_dt <- gwas_dt[, length_vsbid:=lapply(gwas_dt$variant_sbid, length)]

# create a single column data table with as many rows as there're variant sbids
extra_rows <- data.table(rep(gwas_dt$rowID, gwas_dt$length_vsbid))
names(extra_rows) <- c("rowID")

# merge the extra rows and the gwas data table to add more rows
expand_gwas <- merge(gwas_dt, extra_rows, all=T)

# unlist the variant_sbid column to make it into a vector of single variant_sbid values
varsbids <- unlist(gwas_dt$variant_sbid)
setnames(expand_gwas, "variant_sbid", "variant_sbid_list")

# add the expanded varsbids to the expanded data table
expand_gwas[, 'variant_sbid'] <- varsbids

# merge clinvar and the expanded gwas
clinvar_gwas <- merge(clinvar_genes_dt, expand_gwas, by='variant_sbid')
```

We now have fields from both ClinVar and GWAS in the same data table, keyed by `variant_sbid`:

```{r}
colnames(clinvar_gwas)
```

We can look at summary tables for fields, eg:

```{r}
table(unlist(clinvar_gwas$clinical_significance))
```

Alternately, you can extract summary values for fields using SolveBio's facets:

```{r}
ClinVar_variants <- Dataset.get_by_full_path("solveBio:public:/ClinVar/3.7.4-2017-01-30/Variants-GRCh37")
facets <- Dataset.facets(id = ClinVar_variants$id, 
                        filters = genes_filter, 
                        facets = list("clinical_significance", "gene_symbol", "phenotype"))
facets$clinical_significance
facets$gene_symbol
facets$phenotype
```

Let's imagine we have expression data for our gene list of interest for 5 conditions and we want to visualize it with a heatmap:

```{r, message = FALSE}
# simulate an expression matrix for the genes based on min/max values from a real expression dataset:
expr_m <- matrix(runif(5*length(genes),0,20), nrow=length(genes), ncol=5)
expr_m <- expr_m[order(expr_m[,1]),]
row.names(expr_m) <- genes
expr_heatmap <- heatmap(expr_m, 
                        Rowv = NA, 
                        Colv = NA, 
                        col = cm.colors(256), 
                        scale = "column", 
                        margins = c(5,10))
```

Similarly, we can use ExpressionSet from Bioconductor/Biobase package to create the heatmap:

```{r, message = FALSE}
# using expression set heatmap
library("Biobase")
minimalSet <- ExpressionSet(assayData = expr_m)
heatmap(exprs(minimalSet))
```
